server/server.go

```diff
@@ -1854,101 +1934,143 @@ func (s *StanServer) storeIOLoop() {
 	// This will be done by a master election within the cluster, for now we
 	// assume we are the master and assign the sequence ID here.
 	////////////////////////////////////////////////////////////////////////////
-	var storesToFlush map[*stores.ChannelStore]struct{}
-
-	var _pendingMsgs [ioChannelSize]*ioPendingMsg
-	var pendingMsgs = _pendingMsgs[:0]
-
-	storeIOPendingMsg := func(iopm *ioPendingMsg) {
-		cs, err := s.assignAndStore(&iopm.pm)
-		if err != nil {
-			Errorf("STAN: [Client:%s] Error processing message for subject %q: %v", iopm.pm.ClientID, iopm.m.Subject, err)
-			s.sendPublishErr(iopm.m.Reply, iopm.pm.Guid, err)
-		} else {
-			pendingMsgs = append(pendingMsgs, iopm)
-			storesToFlush[cs] = struct{}{}
-		}
-	}
+	storesToFlush := make(map[*stores.ChannelStore]struct{}, 64)
 
 	batchSize := s.opts.IOBatchSize
 	sleepTime := s.opts.IOSleepTime
 	sleepDur := time.Duration(sleepTime) * time.Microsecond
 	max := 0
+	pubMsgs := ioProtoList{}
 
+	ready.Done()
 	for {
 		select {
-		case iopm := <-s.ioChannel:
-			// Create a new map (probably faster than deleting elements down below)
-			storesToFlush = make(map[*stores.ChannelStore]struct{})
-
-			// store the one we just pulled
-			storeIOPendingMsg(iopm)
-
-			remaining := batchSize - 1
-			// fill the pending messages slice with at most our batch size,
-			// unless the channel is empty.
-			for remaining > 0 {
-				ioChanLen := len(s.ioChannel)
-
-				// if we are empty, wait, check again, and break if nothing.
-				// While this adds some latency, it optimizes batching.
-				if ioChanLen == 0 {
-					if sleepTime > 0 {
-						time.Sleep(sleepDur)
-						ioChanLen = len(s.ioChannel)
-						if ioChanLen == 0 {
-							break
-						}
-					} else {
-						break
-					}
-				}
-
-				// stick to our buffer size
-				if ioChanLen > remaining {
-					ioChanLen = remaining
+		case <-s.ioSignal:
+			// If allowed to sleep, to try gather more
+			if sleepTime > 0 {
+				s.ioMutex.RLock()
+				count := s.ioList.count
+				s.ioMutex.RUnlock()
+				if count > 0 && count < batchSize {
+					time.Sleep(sleepDur)
 				}
+			}
+			// Take the list now.
+			s.ioMutex.Lock()
```

Comment:
For readability, this could be a function on the ioProtoList, e.g. ioProtoList.Clear()